Week4

hyeyeonie opened the pull request:
## 📝 작업 내용

### 🔐 Network 회원가입 (POST) - `RegisterViewController` 화면 구현

> 회원가입 화면의 UI 구성 및 `RegisterService`를 통한 POST 네트워크 요청을 처리하였습니다.  
> `NetworkError`에 따른 에러 처리 및 응답 디코딩을 통해 네트워크 통신을 구현하였습니다.

#### 1️⃣ `RegisterService` 싱글톤 클래스 구현
- `makeRequestBody()`에서 JSON 인코딩을 통해 요청 바디 생성  
- `makeRequest()`를 통해 URL, HTTP Method, Header 등 설정  
- `PostRegisterData()`에서 실제 POST 요청 처리 및 응답 디코딩

#### 2️⃣ `RegisterRequestBody`, `RegisterResponseBody` 모델 생성
- Codable 기반으로 POST 요청과 응답에 사용될 데이터 모델 정의

#### 3️⃣ `NetworkError` 열거형 정의
- 인코딩, 응답 에러, 디코딩 실패 등 다양한 네트워크 오류 케이스 처리

#### 4️⃣ `RegisterViewController` UI 구성 및 연결
- 사용자 입력을 위한 `UITextField`와 버튼 구성  
- 버튼 액션에서 `RegisterService` 호출 및 응답 결과 표시 처리

<br>

---

### 🔍 Network 닉네임 조회 (GET) - `InfoViewController` 화면 구현

> 닉네임 키워드를 기반으로 검색된 닉네임 리스트를 조회하고,  
> 결과를 UI에 출력하는 GET 네트워크 요청 및 뷰 컨트롤러를 구현하였습니다.

#### 1️⃣ `InfoViewController` UI 구성
- `UITextField`, `UIButton`, `UILabel`을 `UIStackView`로 정렬  
- 키워드 입력값 실시간 반영 및 버튼 클릭 시 요청 처리

#### 2️⃣ `GetInfoService` 클래스 구현
- `fetchNickNameList(keyword:)` 메서드를 통해 닉네임 리스트 GET 요청  
- 키워드가 비어 있는 경우 전체 리스트 요청 처리

#### 3️⃣ 닉네임 조회 응답 모델 정의
- `CheckNicknameResponseWrapper`, `CheckNicknameResponseBody`로 기능 중심 네이밍 리팩토링  
- 응답 디코딩 및 `infoLabel`에 결과 문자열로 출력

#### 4️⃣ `UITextField` 편집 이벤트 및 버튼 액션 처리
- 키워드 입력값 변경 시 상태 업데이트  
- 검색 버튼 클릭 시 `Task`를 통해 비동기 요청 처리

<br>

---

### ✅ Network 로그인 (POST) - `RegisterViewController` 기능 구현

> 사용자 계정 정보를 입력받아 로그인 요청을 보내고,  
> 성공 시 토큰 등의 인증 정보를 처리하는 로그인 기능을 구현하였습니다.

#### 1️⃣ `LoginService` 싱글톤 클래스 구현
- `makeRequestBody()`를 통해 JSON 인코딩된 요청 바디 생성  
- `makeRequest()`에서 URL, HTTP Method, Header 등 설정  
- `postLoginData()`에서 실제 로그인 요청 및 응답 디코딩 처리

#### 2️⃣ `LoginRequestBody`, `LoginResponseBody` 모델 생성
- Codable 기반으로 로그인 요청/응답에 사용될 모델 정의

#### 3️⃣ `NetworkError`를 활용한 에러 처리
- 요청 실패, 응답 오류, 디코딩 오류 등 다양한 케이스 구분 처리

#### 4️⃣ `RegisterViewController` UI 구성 및 연결
- 사용자 입력을 위한 `UITextField`와 버튼 구성  
- 버튼 액션에서 `LoginService` 호출 및 응답 결과에 따른 화면 처리

<br>

---

### 👤 Network 내 닉네임 조회 (GET) - `SuccessLoginViewController` 화면 구현

> 로그인된 사용자의 닉네임 정보를 GET 요청을 통해 받아오고,  
> UI에 출력하는 화면을 구현하였습니다.

#### 1️⃣ `SuccessLoginViewController ` UI 구성
- `UILabel`, `UIButton` 구성  
- 닉네임 조회 버튼 클릭 시 요청 처리

#### 2️⃣ `MyInfoService` 클래스 구현
- `fetchMyNickname()` 메서드를 통해 내 닉네임 GET 요청  
- Authorization 헤더를 포함하여 인증 처리

#### 3️⃣ 닉네임 응답 모델 정의
- `MyNicknameResponse` 모델로 닉네임 정보 디코딩  
- 추후 확장을 고려한 네이밍 및 구조 설계

#### 4️⃣ 버튼 액션 및 결과 처리
- 버튼 클릭 시 `Task`로 비동기 요청 처리  
- 닉네임 조회 성공 시 `UILabel`에 표시

<br>

---

### ✏️ Network 닉네임 수정 (PATCH) - `SuccessLoginViewController` 화면 구현

> 사용자가 입력한 유저 ID와 닉네임을 PATCH 요청으로 서버에 전달하여  
> 닉네임을 수정하는 기능을 구현하였습니다.

#### 1️⃣ `SuccessLoginViewController` UI 구성
- `UILabel`, `UIButton`으로 구성된 간결한 인터페이스
- SnapKit을 활용한 레이아웃 설정
- 버튼 2개: **내 닉네임 조회**, **내 닉네임 수정**

#### 2️⃣ 닉네임 수정 액션 처리
- "내 닉네임 수정" 버튼 클릭 시 `UIAlertController`로 입력 받기
  - 유저 ID (`UITextField`)
  - 새 닉네임 (`UITextField`)
- 입력값 검증 후 `ModifyService` 호출
- 결과에 따라 `UIAlertController`로 결과 알림 출력

#### 3️⃣ `ModifyService` 클래스
- 싱글톤 패턴(`shared`)으로 구성
- `makeRequest(nickname:keyword:userId:)`로 PATCH 요청 구성
  - URL, HTTP Method, Headers (`Content-Type`, `userId`) 설정
  - `JSONEncoder`를 이용한 body 인코딩
- `modifyNickname(...)` 메서드에서 네트워크 통신 및 예외 처리

#### 4️⃣ 닉네임 요청/응답 모델 정의
- `ModifyRequestBody`: Codable 구조체로 JSON body 구성
- HTTP 응답 코드 확인 → 성공/실패 분기 처리

<br>
<br>

## 📸 스크린샷

<img src="https://github.com/user-attachments/assets/149108bb-beb6-475d-a11c-249303b702f1" width="280">

hyeyeonie committed:
Add: 초기 세팅 및 .gitignore 추가

hyeyeonie committed:
feat: NSAppTransportSecurity 설정 추가로 모든 HTTP 요청 허용

hyeyeonie committed:
feat: LoginViewController UI 레이아웃 설정

- SnapKit, Then 의존성 추가
- 로그인 화면 레이아웃 구성

hyeyeonie committed:
feat: 로그인 화면 UI 구성, 네트워크 POST 요청 및 에러 처리 구현

- RegisterService 클래스 싱글톤 패턴으로 구현
- POST 요청을 위한 Request Body 생성 및 URLRequest 설정
- 회원가입 API 호출을 위한 POST 요청 구현
- NetworkError 처리 추가 및 HTTP 응답 상태 코드 검사
- RegisterResponseBody 디코딩 처리 추가

hyeyeonie committed:
feat: 닉네임 리스트 조회 기능 구현

- InfoViewController에서 닉네임 검색 UI 구성 및 입력값 처리
- GetInfoService를 통해 닉네임 리스트 GET 요청 수행
- 조회 결과를 infoLabel에 출력, 실패 시 에러 메시지 표시
- UITextField 편집 이벤트 처리 및 검색 버튼 액션 연결

hyeyeonie committed:
Add: UITextField Padding extension 추가

hyeyeonie committed:
feat: 로그인 POST 요청 처리 및 SuccessLoginViewController로 전환

hyeyeonie committed:
fix: LoginViewController를 ResterViewController로 추가

hyeyeonie committed:
feat: 내 닉네임 조회 GET 기능 구현

- alert로 userId 입력해 확인 가능

hyeyeonie committed:
feat: 내 닉네임 수정 PATCH 기능 구현

hyeyeonie committed:
Add: 누락된 코드 추가

sem-git commented on practice_week4_Network/practice_week4_Network/Network/GetInfoServiece.swift:
```diff
@@ -0,0 +1,50 @@
+
+import Foundation
+
+final class GetInfoServiece {
```
오타가 있어요 !!

sem-git commented on practice_week4_Network/practice_week4_Network/Extension/UIStackView+.swift:
오 ~~

sem-git commented on practice_week4_Network/practice_week4_Network/Network/DTO/MyInfoResponseBody.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by 신혜연 on 5/13/25.
+//
+
+struct MyInfoResponseBody: Codable {
+    let data: NicknameData
+}
```
혹시 이 ResponseBody에는
```
let success: Bool
let code: String
let message: String
```

를 안 넣어준 이유가 따로 있을까요?

hyeyeonie replied:
아하 ...ᐟ 닉네임 조회만 하면 된다고 생각해서 명세서에 있는 내용을 누락시켰네요 ...ᐟ 수정하겠습니다:)

sem-git commented on practice_week4_Network/practice_week4_Network/Network/GetMyInfoService.swift:
```diff
@@ -0,0 +1,54 @@
+    static let shared = GetMyInfoService()
+    private init() {}
+    
+    func makeRequest(keyword: String?, userId: Int64) -> URLRequest? {
+        var urlString = "http://api.atsopt-seminar4.site/api/v1/users/me"
+        
+        if let keyword = keyword, !keyword.isEmpty,
+           let encoded = keyword.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
+            urlString += "?keyword=\(encoded)"
+        }
+        
+        guard let url = URL(string: urlString) else { return nil }
+        
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.setValue(String(userId), forHTTPHeaderField: "userId")
+        return request
+    }
+    
+    func fetchNickNameList(keyword: String?, userId: Int64) async throws -> String? {
+        guard let request = makeRequest(keyword: keyword, userId: userId) else {
```
내 닉네임 조회에는 keyword 파라미터가 사용되지 않아서, 관련 로직은 지우는 게 더 좋을 것 같아요!
(닉네임 리스트 조회처럼 keyword로 검색하는 것이 아니고 그냥 내 닉네임을 userId로 조회하는 것이기 때문에)

hyeyeonie replied:
아 이해했습니다. keyword에서의 코드가 리스트 조회 시처럼 검색할 때 사용되는 부분이었군요.ᐟ.ᐟ 
수정하겠습니다.ᐟ

sem-git commented on practice_week4_Network/practice_week4_Network/Network/ModifyService.swift:
```diff
@@ -0,0 +1,55 @@
+    static let shared = ModifyService()
+    private init() {}
+    
+    func makeRequest(nickname: String, keyword: String?, userId: Int64) -> URLRequest? {
+        var urlString = "http://api.atsopt-seminar4.site/api/v1/users"
+        
+        if let keyword = keyword, !keyword.isEmpty,
+           let encoded = keyword.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
+            urlString += "?keyword=\(encoded)"
+        }
```
내 닉네임 조회에도 말씀드린 것처럼 불필요한 로직을 걷어내면 더 좋을 것 같네요 ㅎㅎ

sem-git reviewed (commented):
수고 정말 많으셨습니다 ~~
합세도 파이팅합시다 :>

sem-git commented on practice_week4_Network/practice_week4_Network/View/InfoViewController.swift:
```diff
@@ -0,0 +1,96 @@
+            $0.bottom.equalTo(self.view.safeAreaLayoutGuide).inset(170)
+        }
+
+        [keywordTextField, infoLabel, searchButton].forEach {
+            self.stackView.addArrangedSubview($0)
+        }
```
UIStackView+ Extension 잘 만들어놓고 깜빡했나보네요 ㅎㅎ

hyeyeonie committed:
Refactor: 클래스 오타 수정

hyeyeonie committed:
Refactor: 서버 응답 구조에 맞춰 MyInfoResponseBody에 success, code, message 필드 추가

hyeyeonie committed:
Refactor: 기존에 만들어놓은 extension으로 사용

hyeyeonie committed:
Refactor: 불필요한 로직 걷어내기

- keyword 문자열을 URL에 안전하게 넣을 수 있도록 인코딩하는 부분
- 검색어를 문자열로 찾을 때 사용하는 코드
- Int타입의 userId로 찾는 경우에는 불필요한 코드
